# Confidence Interval for Monte Carlo

### A simple example 
(slightly modified from the lecture notes).

In [2]:
"""
Approximate pi by Monte Carlo.  
input:  
    n       integer, number of MC samples
output:  
    pi_mc   float, estimated value of pi.
"""
function approx_pi(n)
    x = 2rand(n) .- 1       # Coordinates
    y = 2rand(n) .- 1

    # Determine if points are inside the circle (a "hit")
    hits = 0
    for i = 1:n
        if x[i]^2 + y[i]^2 <= 1
            hits += 1
        end
    end
    
    pi_mc = 4hits / n

    return(pi_mc)
end

approx_pi

In [23]:
# set random seed for reproducibility
using Random
Random.seed!(1)

# run Monte Carlo (n = 100) for m = 50 times
m = 50
MC_results = [approx_pi(100) for i in 1:m]

# print out results in a nice form
using Printf
for line in 1:5
    for word in 1:10
        @printf("%.2f ", MC_results[(line-1)*10 + word])
    end
    print("\n")
end

print("\nmax = ", maximum(MC_results), ", min = ",  minimum(MC_results))

3.32 2.80 3.00 2.92 3.16 3.32 2.84 3.12 3.56 3.32 
2.92 2.92 3.00 3.12 3.20 3.00 3.32 3.40 3.12 2.96 
2.92 3.16 3.36 2.84 2.88 2.84 3.24 3.08 2.96 3.12 
3.32 3.16 3.32 3.00 3.24 3.08 2.96 2.96 3.16 2.80 
3.04 3.44 3.16 3.28 3.08 3.04 3.16 3.16 3.28 3.24 

max = 3.56, min = 2.8

What do you see? Do you trust these Monte Carlo results? If so, how confident are you?

Notice the maximum of the results is 3.56 and the minimum is 2.8. Maybe a better way to interprete the MC results is to say "$\pi$ is in the interval [2.8, 3.56]"?

The idea here is similar to a shotgun. An individual bullet of a shotgun has low accuracy. But by firing many bullets at once, we are more confident that the target would be in the range of our bullets (estimate by an interval).

### Confidence interval

We say $[a, b]$ ($a, b$ are random numbers) is a 95% confidence interval for $\theta$ if with probability at least 95%, $[a, b]$ covers $\theta$.

If we can obtain a 95% confidence interval for $\pi$ instead of a single approximation, we would have a better understanding of the error of our approximation.

#### One way to construct an (empirical) confidence interval

- Suppose you want to estimate $\theta$ by Monte Carlo with $n$ samples using function MC(n), i.e. $\theta \approx $ MC(n).

- Run MC(n) $m$ times with $m$ being large (at least 10) and get a vector `MC_results` of size $m$ storing the results.

- The final estimation is `est = mean(MC_results)`; the confidence interval is  
`[est - half_range, est + half_range]`, where `half_range = 2 * std(MC_results) / sqrt(m)`.

If you are curious why this would give you a 95% confidence interval, check the Central Limit Theorem and the quantiles of Gaussian distribution.

Example:

In [31]:
# set random seed for reproducibility
using Random
Random.seed!(1)

# run Monte Carlo for m = 50 times
m = 50
MC_results = [approx_pi(100) for i in 1:m]

# compute required statistics
using Statistics
est = mean(MC_results)
half_range = 2 * std(MC_results) / sqrt(m)

using Printf
@printf("estimation of pi is %.2f\n", est)
@printf("95%% confidence interval is [%.2f, %.2f]", est-half_range, est+half_range)

estimation of pi is 3.11
95% confidence interval is [3.06, 3.16]

**The proficient way to do Monte Carlo is to always report a confidence interval for your estimation.**

# Fancier plots

See one example below. Explore [more examples](http://docs.juliaplots.org/latest/generated/gr/) yourself if interested.

In [1]:
using Plots

plotly()

x = 0:0.1:2*π
y = 0:0.1:2*π

# (xx[i,j], yy[i,j]) = (x[i], y[j]), for the purpose of vectorized operation
xx = repeat(x, 1, length(y))
yy = transpose(repeat(y, 1, length(x)))

zz = sin.(xx).*cos.(yy)
plot3d(xx, yy, zz, label=:none, st = :surface)
plot!(xlab="x", ylab="y", zlab="sin(x)*cos(y)")

┌ Info: For saving to png with the Plotly backend PlotlyBase has to be installed.
└ @ Plots /opt/julia/pkg/packages/Plots/oZheM/src/backends.jl:372


<!DOCTYPE html>
 
 
 Plots.jl